### PR TITLE
Fix validate coupon sale items

### LINKED ISSUE #21226: Bug in sale coupon restriction logic following the changes in #21219
Howdy!

WooCommerce will no longer, following the changes here (https://github.com/woocommerce/woocommerce/pull/21219) allow the use of the coupon with the item that is not on sale. Instead, it will remove the coupon and state the coupon cannot be used on a sale item even when one of the items in the cart is not a sale item

The issue initially surfaced here: https://wordpress.org/support/topic/current-release-3-4-5-breaks-coupons/#post-10639339

The user dug into this and found that the issue is with the `foreach` in `validate_coupon_sale_items` function, as we are not checking all items in the cart — as soon as we see an on-sale item, then we exit ( `break;` )

The code in question is in here: 

https://github.com/woocommerce/woocommerce/blob/76312d546f1acbf9caa0e9df55e9528f479dcc17/includes/class-wc-discounts.php#L766-L783

Cheers!

### PR DESCRIPTION
**Pull request by coenjacobs**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Pull request #21219 incorrectly fixed the logic of validating coupons being applied based on on-sale items in cart. It resulted in "per product"-type coupons that were restricted to be used on sale items, would not be allowed to be used in a cart consisting of some products being on-sale and others not. This pull request checks the provided items for both coupons that apply on the whole cart and for those that apply on per product basis. It does that so by reversing the variable it is using to match the expected behavior and returning the opposite of it. This resulted in maintaining most of the existing logic, it just now works two ways. 😄 

Closes #21226

### How to test the changes in this Pull Request:

Automated: 
1. Run test suite, the test was actually good but the underlying code wasn't right.

Manual:
1. Create "Fixed product discount" coupon that is restricted to not apply to on sale items
2. Add a on-sale item and a not in sale item to cart
3. Try to apply the created coupon in 1. to cart - this _will_ work now, while in 3.4.4 it doesn't

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### Changelog entry

> Fixed coupons validation logic bug introduced in 3.4.4 that no longer applied per product coupons that were restricted to not applying on sale products, while there were other products in cart which weren't on sale

### CODE REVIEW

**Comment by claudiulodro** [comment]
Sorry, but I don't see the logic error in [the original version](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-discounts.php#L766):
```
if ( $coupon->get_exclude_sale_items() && 'fixed_product' !== $coupon->get_discount_type() ) {
...
}
return true;
```
The per product coupons have the 'fixed_product' discount type so the function will always return true for those.

**Comment by coenjacobs** [comment]
I've been diving deeper in the logic surrounding coupon code validation and the user reported issue here on GitHub and on .org, but it seems they are related, yet not causing each other. I have yet to figure out why someone's per product coupons stopped applying after 3.4.5, as I don't think this PR _will_ fix that, the fix it brings is still valid in my humble opinion.

The existing logic in itself isn't all bad, but it fails to detect a couple things. The method `validate_coupon_sale_items( $coupon )` is responsible for (docblock):
>Ensure coupon is valid for sale items in the list is valid or throw exception.

But, it doesn't do all the checks right. For example, with the current code on 3.4.4, when applying a sale restricted product discount coupon on a on-sale product, the following exception is thrown:
>Sorry, this coupon is not applicable to selected products.

This is not because the sale item coupon check failed, but because `validate_coupon_excluded_items( $coupon )` flagged it as excluded. I honestly don't know what magic goes into that, but that's beyond the point. The `validate_coupon_sale_items( $coupon )` let it pass, resulting in a different error message than one would expect. After my changes, you will get the correct error message (and it also checks for validation of on-sale items with cart percentage/fixed coupons as well):
>Sorry, this coupon is not valid for sale items.

**Comment by claudiulodro** [comment]
This change seems to introduce a logic error in the validation:

With a cart coupon and no sale items, validation fails.

```
$valid = ( 'fixed_product' !== $coupon->get_discount_type() ); // True when fixed_cart discount.

foreach ( $this->get_items_to_validate() as $item ) {
	if ( $item->product && ! $item->product->is_on_sale() ) { // True when no sale items.
		$valid = ! $valid; // Valid is now false and will throw exception.
		break;
	}
}
```

Looking at the [video the user posted of the error](https://www.dropbox.com/s/vikywyz8945z804/Recording%20%2332.mp4?dl=0) it was a percentage discount they were using and not per product discounts. Unless I'm off base somewhere, it is good that the coupon stopped working after the update. That means that #21062 was fixed.

**Comment by soft79** [comment]
Please note that both 'fixed_product' and 'percent' are product discounts. 'fixed_cart' is the only cart discount in core.

I.e. the same logic that applies to 'fixed_product' should also apply to 'percent'.

**Comment by claudiulodro** [comment]
I'll need to do some more research around the historical behavior, but the documented behavior is:

> Percentage discount – **A percentage discount for the entire cart**. For example, if the cart contains three (3) t-shirts @ $20 each = $60, a coupon for 10% off applies a discount of $6.

> Fixed cart discount – **A fixed total discount for the entire cart**. For example, if the cart contains three (3)  t-shirts @ $20 each = $60, a coupon for $10 off gives a discount of $10.

> Fixed product discount – **A fixed total discount for selected products only**. Customer receives a set amount of discount per item. For example, three (3) t-shirts @ $20 each with a coupon for $10 off applies a discount of $30.

https://docs.woocommerce.com/document/coupon-management/

**Comment by soft79** [comment]
Since 3.0 (2.7) it's a per-product discount. See this excerpt from https://woocommerce.wordpress.com/2016/12/15/woocommerce-2-7-beta-1-is-here/  :

We merged cart percent and product percent coupon types into one and removed product_cart discounts. The discounts these coupons provide are identical, however, the cart based validation would stop the coupon being applied if any non-eligble item was in the cart, rather than just discounting eligible items like product coupons do. This was not intuitive, caused store owner and confusion, and most important of all, just meant users would have to checkout twice to make use of these coupons (thats not fair nor ideal).

**Comment by claudiulodro** [comment]
I see. Thanks for letting me know. I'll get that fixed and make a note to update the doc.

**Comment by soft79** [comment]
FYI; this function from wc-coupon-functions.php should be used to decide whether it's a product coupon:

`function wc_get_product_coupon_types() {
	return (array) apply_filters( 'woocommerce_product_coupon_types', array( 'fixed_product', 'percent' ) );
}`

**Comment by walkinverse** [comment]
The issue with the coupon is that you can add a non sale item and a sale item.  You have a coupon set to percentage and to ignore on sale items and only allow it to be used with 1 item in the cart.  In 3.4.4 no, issue, everything works are expected.  3.4.5 it does not.  Currently (3.4.5) if you try to apply the coupon it kicks out and will not work on the non sale item.  If you remove the sale item and try again the coupon work.  You add the sale item back to the cart with the non- sale item it throws an error saying coupon cannot be used, which is incorrect.  Please see the video. 3.4.4 is normal behavior, 3.4.5 is not.

https://www.dropbox.com/s/vikywyz8945z804/Recording%20%2332.mp4?dl=0

File: includes/class-wc-discounts.php [hunk 1]
```diff
@@ -764,21 +764,23 @@ protected function validate_coupon_product_categories( $coupon ) {
...
+			return true;
+		}
 
-			foreach ( $this->get_items_to_validate() as $item ) {
-				if ( $item->product && $item->product->is_on_sale() ) {
-					$valid = false;
-					break;
-				}
-			}
+		$valid = ( 'fixed_product' !== $coupon->get_discount_type() );
```
Comment (soft79) [inline includes/class-wc-discounts.php]: Replace by:

    $valid = ! $coupon->is_type( wc_get_product_coupon_types() );

**Review by soft79 — CHANGES_REQUESTED** [review]
Instead of 'fixed_product' we must verify all product type discounts ( i.e. wc_get_product_coupon_types() ). See remark in code.

Additionally something that is on my mind:

The logic of validate_coupon_sale_items() in this PR is very different from the current handling of excluded items in WC_Discounts. Exclusions for cart-discount are handled in validate_coupon_eligible_items().

	protected function validate_coupon_eligible_items( $coupon ) {
		if ( ! $coupon->is_type( wc_get_product_coupon_types() ) ) {
			$this->validate_coupon_excluded_product_ids( $coupon );
			$this->validate_coupon_excluded_product_categories( $coupon );
			// Here we should call validate_coupon_excluded_sale_items( .. )
		}

		return true;
	}


Current tests are also lacking some combinations of coupons/cart with/without sale items.

I would be happy to create a PR this weekend; but currently there are already 2 PR's active about this issue...

**Comment by claudiulodro** [comment]
#21241 should solve the issue and prevent it from happening again in the future.

**Comment by soft79** [comment]
That looks better. Cheers!

**Comment by Fit4home** [comment]
Hi there!
Could the problem with the coupons be solved?

**Comment by walkinverse** [comment]
I am still waiting for the fix to be released as I cant use the coupons until its sent out, unfortunately.  I do think they fixed it though.

**Pull request closed without merging**